lib_eio/unix/rcfd.ml

```diff
@@ -0,0 +1,171 @@
+
+   If you increment/decrement when in the state [closing/no-users] or [closing/closed],
+   you do not get/return a share.
+
+   A fiber can call {!close} to move from [open] at any time.
+   If there are operations in progress when it does this, it transitions to [closing/users].
+   Otherwise, it goes directly to [closing/no-users].
+   The fiber that does this is known as the "closing fiber"
+   and is responsible for finishing the close.
+
+   We move from [closing/users] to [closing/no-users] when [ops] becomes 0
+   (and [t] therefore owns all shares of [x]).
+   [ops] may continue to change after this, but we never return to [closing/users]
+   or give any user a share of [x] after this.
+
+   From [closing/no-users], we transition to [closing/closed],
+   transferring the now-full ownership of [x] to the closing fiber.
+
+   In reality, the three [closing/*] states are represented by the single constructor [Closing],
+   and the code must work whatever the true state might be. *)
+
+type state =
+  | Open of Unix.file_descr
+  | Closing of (unit -> unit)     (* Function is called when [ops] becomes 0. *)
+
+type t = {
+  ops : int Atomic.t;
+  fd : state Atomic.t;
+}
+
+let fully_closed = Closing ignore      (* Used for [closing/closed] *)
+
+let put t =
+  let old = Atomic.fetch_and_add t.ops (-1) in
+  if old = 1 then (
+    (* We decremented [ops] from one to zero. We may need to notify the closer. *)
+    match Atomic.get t.fd with
+    | Open _ -> ()     (* The fast path. We're not closing. *)
+    | Closing no_users as prev ->
+      (* There are four possibilities for the state when we did the decrement:
+         - open: But it got closed after that. There could be new active users by now.
+         - closing/users: We were the last user and transitioned to closing/no-users.
+                          We need to notify the closer, or make sure someone else will do it later.
+         - closing/no-users: We might need to notify, since a previous thread that reached zero
+                             might have then seen [ops > 0] and deferred it to us.
+         - closing/closed: No need to do anything, but notifying is harmless.
+       *)
+      if Atomic.get t.ops > 0 then ()  (* Someone else will deal with it. *)
+      else if Atomic.compare_and_set t.fd prev fully_closed then (
+        (* We observed [t.ops = 0] after closing, so we were then at either [closing/no-users]
+           or [closing/closed], and we're now certainly at [closing/closed].
+           If it was [closing/no-users] then we now own the FD, which we pass to the closer.
+           If it was [closing/closed] then we don't, but [no_users] is [ignore] anyway. *)
+        no_users ()
+      ) else (
+        (* Someone else notified the closer first. We're now in [closing/closed]. *)
+      )
+  ) else (
+    assert (old > 1)
+  )
+
+let get t =
+  Atomic.incr t.ops;
```

Comment:
Domain B gets here, meanwhile Domain A already decided to close with `no_users`.
Domain B sees `Closing` since the atomics are ordered, it aborts by doing `put` on line 97.
B sees the same sequence of events of A, and ends up calling `no_users` as well, so you end up with a double close. The assumption is that the `no_close` of A would be carried over to B, which I think doesn't hold here, both end up calling it.

Reply:
`put` always does a CAS to `` Closing ignore`` before calling `no_users`, so the real function can only get called once.

Reply:
ahA! indeed, I've missed the ignore in `fully_closed`